### PR TITLE
[WIP] Support switching heap manager at runtime

### PR DESCRIPTION
**Pull request by Celthi**
Hi @yanqi27 ,
This PR is to support switching heap manager at runtime. I haven't finished all the code yet. But I'll share it in the front in case the direction is not good to go.
My approach here is defined a struct (I cannot use class as it is C) to store the heap related function, and call them by `CA_HEAP->func()`. The CA_HEAP is a macro that points to the current heap manager.
A user can switch to another heap by typing the command `switch_heap <heap manager name>` while inside the gdb.

Could you review this approach? It can compile and I will fix all the other places later if we decide to go this way.

```c
struct CoreAnalyzerHeapInterface {
    HeapVersionFunc heap_version;
    InitHeapFunc init_heap;
    HeapWalkFunc heap_walk;
    IsHeapBlockFunc is_heap_block;
    GetHeapBlockInfoFunc get_heap_block_info;
    GetNextHeapBlockFunc get_next_heap_block;
    GetBiggestBlocksFunc get_biggest_blocks;
    /*
    * Get all in-use memory blocks
    * 	If param opBlocks is NULL, return number of in-use only,
    * 	otherwise, populate the array with all in-use block info
    */
    WalkInuseBlocksFunc walk_inuse_blocks;

};
```

### CODE REVIEW

**Comment by Celthi** [comment]
There are some commits that have been merged before. I have no idea why they will show up. I will remove them and solve the conflict later.

**Review by yanqi27 — COMMENTED** [review]
It looks good. I like the interface design. It is easy to maintain and expand.
Do you have a command to list all supported heap types? Another possibility is search key variables in the target process to determine the heap manger used. Then we can automatically choose for the user.

File: src/heap_ptmalloc.cpp [hunk 1]
```diff
@@ -2185,3 +2199,7 @@ memcpy_field_value(struct value *val, const char *fieldname, char *buf,
 	}
 	return true;
 }
+
+class PtMallocHeap: public CoreAnalyzerHeapInterface {
```
Comment (Celthi) [inline src/heap_ptmalloc.cpp:2203]: This line will be removed. In the beginning, I would like to use the c++ class to define the interface. Later I thought the gdb is written only in C and the class cannot be used. So I switch to using the C struct.  So the current version is using the C struct. This line is obsolete. But it turns out the gdb can use c++ code even if the file is with `.c` suffix. It is kind of strange, I haven't fully understood how it works, so I will stick to the C struct interface.
cc @yanqi27

Comment (yanqi27) [reply src/heap_ptmalloc.cpp:2203]: Probably around gdb v9, it switches to c++ compiler. It probably compiles all sources with c++ regardless to its suffix.

**Comment by Celthi** [comment]
I'm thinking it might be better that each heap manager (different version is treated as a different heap manager) provides the same interface.
For example, pt 2.12 is a different heap manager than pt 2.31 though they are different versions. The common code between them can be extracted into a commonplace. They all called the common code.

I'm planing to do this refactor after this PR settled down.

cc @yanqi27

File: src/heap_ptmalloc.cpp (hunk 1 above)
Comment (Celthi) [reply src/heap_ptmalloc.cpp:2203]: If cpp could be used, I'm thinking switching to use C++ abstract class to define the interface.

**Comment by yanqi27** [comment]
It is a smart approach to support multiple versions of the same heap manager. It also shows the power of the interface.

**Comment by Celthi** [comment]
Hi @yanqi27 Could you review this PR? It is ready to be reviewed. I have run the test for pt malloc and it is passed. For the tc malloc, I modify the test's makefile so the test config is right now. But the tc malloc has a error, i think it is related to my tc malloc is either too new or too old and it is not related to my change.

```
vscode ➜ /workspaces/core_analyzer/test (master ✗) $ make check-tcmalloc
g++ -g -m64 -fpermissive -DTCMALLOC_TEST -c -o mallocTest_tcmalloc.o mallocTest.cpp
g++ -g -m64 -fpermissive -o mallocTest_tcmalloc mallocTest_tcmalloc.o -lpthread  -ltcmalloc
/workspaces/core_analyzer/build/gdb-9.2/build/gdb/gdb2 --data-directory=/usr/share/gdb mallocTest_tcmalloc -q -ex 'switch_heap tc' -x verify.py
Reading symbols from mallocTest_tcmalloc...
Swith to use tc malloc.No registers.
[ca_test] ==== Test Against Live Process ====
Breakpoint 1 at 0x257f: file mallocTest.cpp, line 125.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff781c700 (LWP 68040)]
[New Thread 0x7ffff701b700 (LWP 68041)]

Thread 1 "mallocTest_tcma" hit Breakpoint 1, last_call () at mallocTest.cpp:125
125     {
[ca_test] Checking heap blocks ...
Failed to lookup gv "kPageShift"
tcmalloc heap was not initialized successfully
[ca_test] Failed to query block at 0x555555dae000
[ca_test] Test failed
```


The pt malloc passed as follows
```
vscode ➜ /workspaces/core_analyzer/test (master ✗) $ make check
/workspaces/core_analyzer/build/gdb-9.2/build/gdb/gdb2 --data-directory=/usr/share/gdb mallocTest -q -x verify.py
Reading symbols from mallocTest...
[ca_test] ==== Test Against Live Process ====
Breakpoint 1 at 0x244e: file mallocTest.cpp, line 126.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff7a29700 (LWP 70670)]
[New Thread 0x7ffff7228700 (LWP 70671)]

Thread 1 "mallocTest" hit Breakpoint 1, last_call () at mallocTest.cpp:126
126             std::cout << "This is the last function call\n";
[ca_test] Checking heap blocks ...
[ca_test]       Verified 8209 heap blocks
[ca_test] Checking biggest heap blocks ...
[ca_test]       Found top 16 biggest heap blocks
[ca_test] Checking heap walk ...
[ca_test]       Heapwalk discovered 4224 in-use blocks and 2905 free blocks
[ca_test] Checking C++ objects ...
[ca_test]       Found 4 "Derived" objects
[ca_test] Checking object reference ...
[ca_test]       Found the global reference: var "hidden_object" at 0x55555555d2e0
[ca_test]       Found heap reference: addr=0x55555556ffd0 size=24 to object at address 0x55555556ffb0
[ca_test] ==== Test Against Core Dump ====
Saved corefile core.70662
[Inferior 1 (process 70662) killed]
[New LWP 70662]
[New LWP 70670]
[New LWP 70671]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/workspaces/core_analyzer/test/mallocTest'.
Program terminated with signal SIGTRAP, Trace/breakpoint trap.
#0  last_call () at mallocTest.cpp:126
126             std::cout << "This is the last function call\n";
[Current thread is 1 (Thread 0x7ffff7a5b740 (LWP 70662))]
[ca_test] Checking heap blocks ...
Could not open /proc/70662/maps
Target process has changed. Rebuild heap information
[ca_test]       Verified 8209 heap blocks
[ca_test] Checking biggest heap blocks ...
--Type <RET> for more, q to quit, c to continue without paging--
[ca_test]       Found top 16 biggest heap blocks
[ca_test] Checking heap walk ...
[ca_test]       Heapwalk discovered 4224 in-use blocks and 2905 free blocks
[ca_test] Checking C++ objects ...
[ca_test]       Found 4 "Derived" objects
[ca_test] Checking object reference ...
[ca_test]       Found the global reference: var "hidden_object" at 0x55555555d2e0
[ca_test]       Found heap reference: addr=0x55555556ffd0 size=24 to object at address 0x55555556ffb0
[ca_test] Pass
```

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.205.2/containers/cpp/.devcontainer/base.Dockerfile
+
+# [Choice] Debian / Ubuntu version (use Debian 11/9, Ubuntu 18.04/21.04 on local arm64/Apple Silicon): debian-11, debian-10, debian-9, ubuntu-21.04, ubuntu-20.04, ubuntu-18.04
+ARG VARIANT="bullseye"
```
Comment (Celthi) [inline .devcontainer/Dockerfile:4]: I use this container to develop the code in Mac OS. So I can twist as whatever I like without breaking my host OS.

Comment (yanqi27) [reply .devcontainer/Dockerfile:4]: cool

File: src/heap.cpp [hunk 3]
```diff
@@ -10,6 +10,22 @@
...
+	#ifndef _WIN32
+	// heap.cpp is used by ref.dll, so we need to only support switching heap in Linux
+    gCoreAnalyzerHeaps[HeapManagerPtMalloc] = get_pt_malloc_heap_manager();
+	gCoreAnalyzerHeaps[HeapManagerTcMalloc] = get_tc_malloc_heap_manager();
+	#endif
+	#ifdef _WIN32
+	gCoreAnalyzerHeaps[HeapManagerMscrtMalloc] = get_mscrt_malloc_heap_manager();
+	#endif
+}
+EnumHeapManager gCurrentHeap = HeapManagerPtMalloc;
```
Comment (yanqi27) [inline src/heap.cpp:27]: Could you move this to the top of the file so that it is easier to spot the global vars?

File: src/heap.h [hunk 4]
```diff
@@ -47,11 +28,74 @@ struct inuse_block
...
+ * @brief each heap manager is assigned to an enum
+ * 
  */
-extern bool walk_inuse_blocks(struct inuse_block* opBlocks, unsigned long* opCount);
+enum EnumHeapManager {
+    HeapManagerReserved = 0,
+    HeapManagerPtMalloc = 1,
+    HeapManagerTcMalloc = 2,
+	HeapManagerMscrtMalloc = 3,
+	HeapManagerLastOne,
```
Comment (yanqi27) [inline src/heap.h:40]: Indentation

File: src/heap.h [hunk 5]
```diff
@@ -47,11 +28,74 @@ struct inuse_block
...
+};
+
+typedef const char * (*HeapVersionFunc)(void);
+typedef bool (*InitHeapFunc)(void);
+typedef  bool (*HeapWalkFunc)(address_t addr, bool verbose);
```
Comment (yanqi27) [inline src/heap.h:45]: extra space

**Comment by Celthi** [comment]
Note: I only modified the gdb-9.2. And I think we should have a script that modified the gdb `makefile.in` on the fly, so we can support more than one version of gdb with just one script. For example, the flow I like is 

1. Download the gdb zip from the official website and unzip it.
2. Run a script copy those file into the above repo and modify the `makefile.in`
3. Compile the code.

With the script we don't need to have multiple folders for multiple gdb versions.

File: src/heap_ptmalloc.cpp [hunk 6]
```diff
@@ -2089,6 +2103,25 @@ static bool get_glibc_version(void)
...
+	if (sym == NULL) {
+		CA_PRINT("Failed to lookup gv \"mp_\"\n");
+		return false;
+	}
+	val = value_of_variable(sym, 0);
+	data = (char*)value_as_address(val);
+	
+	const char* version = data;
+
+#endif // Fixme: We should read the libc_version from the debugee, but the above code never works.
```
Comment (yanqi27) [inline src/heap_ptmalloc.cpp:2124]: This is a static var. Maybe `VAR_DOMAIN` is not enough.

**Review by yanqi27 — APPROVED** [review]

File: src/heap_ptmalloc.cpp (hunk 6 above)
Comment (Celthi) [reply src/heap_ptmalloc.cpp:2124]: I tried LABEL_DOMAIN, but it is not work either. I cannot find a suitable domain from
```c

typedef enum domain_enum_tag
{
  /* UNDEF_DOMAIN is used when a domain has not been discovered or
     none of the following apply.  This usually indicates an error either
     in the symbol information or in gdb's handling of symbols.  */

  UNDEF_DOMAIN,

  /* VAR_DOMAIN is the usual domain.  In C, this contains variables,
     function names, typedef names and enum type values.  */

  VAR_DOMAIN,

  /* STRUCT_DOMAIN is used in C to hold struct, union and enum type names.
     Thus, if `struct foo' is used in a C program, it produces a symbol named
     `foo' in the STRUCT_DOMAIN.  */

  STRUCT_DOMAIN,

  /* MODULE_DOMAIN is used in Fortran to hold module type names.  */

  MODULE_DOMAIN,

  /* LABEL_DOMAIN may be used for names of labels (for gotos).  */

  LABEL_DOMAIN,

  /* Fortran common blocks.  Their naming must be separate from VAR_DOMAIN.
     They also always use LOC_COMMON_BLOCK.  */
  COMMON_BLOCK_DOMAIN,

  /* This must remain last.  */
  NR_DOMAINS
} domain_enum;

```

Comment (Celthi) [reply src/heap_ptmalloc.cpp:2124]: solved in https://github.com/yanqi27/core_analyzer/pull/38/files

**Pull request merged**